src/hotspot/cpu/x86/stubGenerator_x86_64_string.cpp

```diff
@@ -0,0 +1,2012 @@
+  bool isUU = (ae == StrIntrinsicNode::UU);
+  bool isU = isUL || isUU;  // At least one is UTF-16
+
+  // Only optimize when haystack can fit on stack with room
+  // left over for page fault prevention
+  assert((COPIED_HAYSTACK_STACK_OFFSET == 0), "Must be zero!");
+  assert((COPIED_HAYSTACK_STACK_SIZE == 64), "Must be 64!");
+
+  // Copy incoming haystack onto stack
+  {
+    Label L_adjustHaystack, L_moreThan16;
+
+    // Copy haystack to stack (haystack <= 32 bytes)
+    __ subptr(rsp, COPIED_HAYSTACK_STACK_SIZE);
+    __ cmpq(haystack_len, isU ? 0x8 : 0x10);
+    __ ja_b(L_moreThan16);
+
+    __ movq(tmp, COPIED_HAYSTACK_STACK_OFFSET + 0x10);
+    __ movdqu(XMM0, Address(haystack, haystack_len, isU ? Address::times_2 : Address::times_1, -0x10));
+    __ movdqu(Address(rsp, COPIED_HAYSTACK_STACK_OFFSET), XMM0);
+    __ jmpb(L_adjustHaystack);
+
+    __ bind(L_moreThan16);
+    __ movq(tmp, COPIED_HAYSTACK_STACK_OFFSET + 0x20);
+    __ vmovdqu(XMM0, Address(haystack, haystack_len, isU ? Address::times_2 : Address::times_1, -0x20));
+    __ vmovdqu(Address(rsp, COPIED_HAYSTACK_STACK_OFFSET), XMM0);
+
+    __ bind(L_adjustHaystack);
+    __ subq(tmp, haystack_len);
+    if (isU) {
+      // For UTF-16, lengths are half
+      __ subq(tmp, haystack_len);
+    }
+    __ leaq(haystack, Address(rsp, tmp, Address::times_1));
+  }
+
+  // Creates a mask of (n - k + 1) ones.  This prevents
+  // recognizing any false-positives past the end of
+  // the valid haystack.
+  __ movq(mask, -1);
+  __ subq(haystack_len, needle_len);
+  __ incrementq(haystack_len);
+  if (isU) {
+    __ shlq(haystack_len, 1);
+  }
+  __ bzhiq(mask, mask, haystack_len);
+
+  for (int size = 1; size <= OPT_NEEDLE_SIZE_MAX; size++) {
+    // Broadcast next needle byte into ymm register
+    int position = isUU ? (size - 1) * 2 : size - 1;
+    if (isUU) {
+      __ vpbroadcastw(XMM0, Address(needle, position), Assembler::AVX_256bit);
+    } else if (isUL) {
+      // Expand needle
+      __ movzbl(rax, Address(needle, position));
+      __ movdl(XMM0, rax);
+      // Byte of needle to words
+      __ vpbroadcastw(XMM0, XMM0, Assembler::AVX_256bit);
+    } else {
+      __ vpbroadcastb(XMM0, Address(needle, position), Assembler::AVX_256bit);
+    }
+
+    // Compare next byte.  Keep the comparison mask in mask, which will
+    // accumulate
+    if (isU) {
+      __ vpcmpeqw(XMM1, XMM0, Address(haystack, position), Assembler::AVX_256bit);
+    } else {
+      __ vpcmpeqb(XMM1, XMM0, Address(haystack, position), Assembler::AVX_256bit);
+    }
+    __ vpmovmskb(tmp, XMM1, Assembler::AVX_256bit);
+    __ andq(mask, tmp);  // Accumulate matched bytes
+    __ testl(mask, mask);
+    __ je(L_noMatch);
+
+    if (size != OPT_NEEDLE_SIZE_MAX) {
+      // Found a match for this needle size
+      __ cmpq(needle_len, size);
+      __ je(L_foundall);
+    }
+  }
+
+  __ bind(L_foundall);
+  __ tzcntl(rax, mask);
+
+  if (isU) {
+    __ shrq(rax, 1);
+  }
+
```

Comment:
We need to be consistent either use tzcntl, shrl, testl or tzcntq, shrq, testq.

Reply:
I'll search through the code making them all consistent.